### PR TITLE
[#8002] InstructorFeedbackResultsPage: add section displaying options for instructors

### LINKED ISSUE #8002: InstructorFeedbackResultsPage: add section displaying options for instructors
As @damithc mentioned in #7815 it will be good to provide the option for question display view that will define behavior during choosing which responses are related to chosen section and which are not.
Here 4 options:
1) Only giver in section -> all responses where giver in chosen section will be included
2) Only recipient in section ->  all responses where recipient in chosen section will be included
3) Both giver and recipient in section -> only responses where both giver and recipient in chosen section will be marked as related to this section
4) Giver or recipient in section -> responses where giver or recipient in chosen section will be displayed

### PR DESCRIPTION
**Pull request by jacoblipech**
Fixes #8002 

**Outline of Solution**
- [x] Add a new dropdown in `filterEditModal.tag` to show 4 options after selecting a `section`
- [x] Make filter edit modal size to be bigger to show all 4 options
- [x] Initialise the sectionDetails in `InstructorFeedbackResultsFilterPanel.java`
- [x] Add a new variable `sectionDetail` to populate the option selected in dropdown as section detail 
- [x] Add a new variable `selectedSectionDetail` to find the option selected by the instructor
- [x] Add a `if showing only a section` field and logic to `filterPanel.tag` to show `selectedSectionDetail` if a section is selected
- [x] Edit styling of `filterPanel.tag` to show the information clearer in the space
- [x] Handle logic to filter according to all 4 options of `selectedSectionDetail` for individual questions loaded by AJAX in web view
- [x] Handle logic to filter according to all 4 options of `selectedSectionDetail` when loading the entire session results in web view
- [x] Modify `InstructorFeedbackDownloadAction` and `InstructorFeedbackDownloadAction` to show the `selectedSectionDetail` when downloading file
- [x] Handle logic to filter response according to `selectedSectionDetail` when downloading the entire csv
- [x] Handle logic to filter responses according to `selectedSectionDetail` when downloading individual questions
- [x] Add the `selectedSectionDetail` field to `copyPasteModal.tag` to handle csvToHtml in download action.
- [x] Add tests for the new feature (currently null to ensure it is running)

**Note:** The current `No Response` field for csv and web view do not tally completely.  I feel that there is no need to focus so much on that as there are only minor differences.

Sample feature gif:
filter session detail ui

Current UI of the feature:
screen shot 2018-07-12 at 12 13 54 am
The default value will be `Show response if either the giver or evaluee is in the selected section`.

screen shot 2018-07-06 at 1 46 05 am
Adjusted the sizing of the filter panel shown

screen shot 2018-07-06 at 1 44 06 am
Shows the `selectedSectionDetail` chosen. Populate the correct data.

### CODE REVIEW

**Comment by jacoblipech** [comment]
@damithc while working on the download results logic for selecting `section` based on `giver` or `recipient`, I realise that some question types has recipient fields as `none` hence a dash.

For example, Mcq type questions, the recipient field is shown as a dash. 
screen shot 2018-07-09 at 1 07 57 pm

However for other question types such as Essay Type, Point Distribution Type and Rank Type questions, the recipient will be the same as giver as shown below.
screen shot 2018-07-07 at 2 30 16 pm

Is there any specific reason for this? Currently, I am not sure if these recipients with dash should be shown or not when people select view by recipients given a `section`.

I am proposing to standardise these questions recipients to be the same as giver instead of a dash. Hence, when people select view by recipients given a `section`, these can be seen as well. What do you think?

**Comment by damithc** [comment]
The recipient depends on the feedback path, not the question type. Any question type can be without a recipient. Similarly, if the feedback path is set as 'self-feedback', the recipient is same as giver.

**Comment by jacoblipech** [comment]
@damithc I see. So in the case above of the `Mcq Question 6`. If I select section as `tutorial group 1` to show `recipient`, should the giver comment by `Alice Betsy: C, Java` be shown or not shown? >.<

**Comment by damithc** [comment]
If the display criteria is 'show if recipient in section' but the response doesn't have a recipient, should we display the response? I think the intuitive answer is 'No'. However, if the criteria is 'show if both in section' and if there is no recipient, we should show as long as the giver is in the section, right? What do you think?

And yes, we should match the web view with the csv file, whichever decision we make for the above.

**Comment by jacoblipech** [comment]
>If the display criteria is 'show if recipient in section' but the response doesn't have a recipient, should we display the response? I think the intuitive answer is 'No'. However, if the criteria is 'show if both in section' and if there is no recipient, we should show as long as the giver is in the section, right? What do you think?
Yes I do agree with this behaviour. I will work towards making the web view, download session and download question to be consistent with this behaviour.

On a side note: I noticed a part of the code in `FeedbackResponsesDb` to populate responses for instructor in the web view where giver is none (dash) and recipient is the selected section. I could not reproduce this situation after configuring the visibility options so I am thinking of removing this part as it may no longer be relevant. @damithc Is there any case of such response/situation happening?

**Comment by damithc** [comment]
> On a side note: I noticed a part of the code in FeedbackResponsesDb to populate responses for instructor in the web view where giver is none (dash) and recipient is the selected section. I could not reproduce this situation after configuring the visibility options so I am thinking of removing this part as it may no longer be relevant. @damithc Is there any case of such response/situation happening?

I can't think of any situation where there is no giver, but there can be cases where the giver is not shown.   Perhaps do a separate PR on removing that code so that it can be discussed/reviewed by others independently?

**Comment by jacoblipech** [comment]
@xpdavid @tshradheya Can help me do an initial review before I move on to the testing? I have tested the results manually for "Session with different question types" in the demo course and it works as required. 

I have also traced `FeedbackSessionLogic` a few times to remove uncalled functions and only add sectionDetail to necessary parts. The commits may not be in order but I have tried to explain the changes.

**Comment by xpdavid** [comment]
I will take a look if I have time as currently, I need to take care of my two GSoCers :P

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 1]
```diff
@@ -1611,6 +1575,42 @@ private FeedbackSessionResultsBundle getFeedbackSessionResultsForUserInSectionBy
                         visibilityTable, responseStatus, roster, responseComments);
     }
 
+    private Map<String, String> initializeParamsWithSelectedSectionDetail(String sectionDetail) {
```
Comment (tshradheya) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: How about using `giver`, `reciever`, `giverOrReciever`, `giverAndReciever` ? Future developers may be confused with `either` and `both`

File: src/main/webapp/WEB-INF/tags/instructor/results/copyPasteModal.tag [hunk 2]
```diff
@@ -7,11 +7,13 @@
 <%@ attribute name="courseId" required="true" %>
 <%@ attribute name="feedbackSession" required="true" %>
 <%@ attribute name="selectedSection" required="true" %>
+<%@ attribute name="selectedSectionDetail" required="true" %>
 <form id="csvToHtmlForm">
   <input type="hidden" name="<%=Const.ParamsNames.COURSE_ID%>" value="${courseId}">
   <input type="hidden" name="<%=Const.ParamsNames.FEEDBACK_SESSION_NAME%>" value="${feedbackSession}">
   <input type="hidden" name="<%=Const.ParamsNames.USER_ID%>" value="${data.account.googleId}">
   <input type="hidden" name="<%=Const.ParamsNames.FEEDBACK_RESULTS_GROUPBYSECTION%>" value="${fn:escapeXml(selectedSection)}">
+  <input type="hidden" name="<%=Const.ParamsNames.FEEDBACK_RESULTS_GROUPBYSECTIONDETAIL%>" value="${fn:escapeXml(selectedSectionDetail)}">
```
Comment (tshradheya) [inline src/main/webapp/WEB-INF/tags/instructor/results/copyPasteModal.tag]: Why do we need need this here?

File: src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag [hunk 3]
```diff
@@ -6,7 +6,7 @@
 <%@ attribute name="filterPanel" type="teammates.ui.template.InstructorFeedbackResultsFilterPanel" required="true" %>
 
 <div id="editModal" class="modal fade" role="dialog">
-  <div class="modal-dialog">
+  <div class="modal-dialog" style="width: 850px">
```
[the remark this answers is not in the document]
Comment (LiHaoTan) [reply src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: Just a note that a fixed width is usually not a good idea, prefer defining them in terms of grid columns.

File: src/main/webapp/WEB-INF/tags/instructor/results/copyPasteModal.tag (hunk 2 above)
Comment (jacoblipech) [reply src/main/webapp/WEB-INF/tags/instructor/results/copyPasteModal.tag]: This is necessary for the `Non-English characters not displayed properly in the downloaded file? Click here.` part in the sessions result page to show the copy paste modal character in the same way as downloaded csv file when filter by a specific section.

**Comment by jacoblipech** [comment]
@damithc while working on the test cases, I noticed a case where the `giverSection` is none while the `recipientSection` is present as shown below.

screen shot 2018-07-25 at 10 09 44 pm

It seems that an instructor can give a response to a specified section of student. As of such cases, I was thinking of making this response show when it is filtered by `either giver or evaluee`, `evaluee` and `both giver and evaluee` as section 1

**Comment by damithc** [comment]
> It seems that an instructor can give a response to a specified section of student. As of such cases, I was thinking of making this response show when it is filtered by either giver or evaluee, evaluee and both giver and evaluee as section 1

Sounds good.

**Comment by jacoblipech** [comment]
@tshradheya @xpdavid Added the necessary tests for this. Ready for review

File: src/main/java/teammates/logic/api/Logic.java [hunk 4]
```diff
@@ -1798,7 +1780,7 @@ public FeedbackSessionResultsBundle getFeedbackSessionResultsForInstructorFromQu
```
Comment (tshradheya) [inline src/main/java/teammates/logic/api/Logic.java]: Add `Assumption.assertNotNull(selectedSectionDetail);`

File: src/test/java/teammates/test/cases/logic/FeedbackQuestionsLogicTest.java [hunk 5]
```diff
@@ -68,12 +68,12 @@ private void testGetRecipientsForQuestion() throws Exception {
         recipients = fqLogic.getRecipientsForQuestion(question, email);
         assertEquals(recipients.size(), 2); // 3 - giver = 2
 
-        ______TS("empty case: response to team members, but alone");
+        ______TS("response to team members, total 2");
```
Comment (tshradheya) [inline src/test/java/teammates/test/cases/logic/FeedbackQuestionsLogicTest.java]: Technically we are missing out on a boundary value test case. Any way to avoid this change?

Comment (jacoblipech) [reply src/test/java/teammates/test/cases/logic/FeedbackQuestionsLogicTest.java]: I have changed student 2 from another section to this section, causing this change. I found that there is a course 3 with only 1 student enrolled, so I can use student in course 3 to test the boundary case!

Comment (tshradheya) [reply src/test/java/teammates/test/cases/logic/FeedbackQuestionsLogicTest.java]: > so I can use student in course 3 to test the boundary case!

Yes. Add back the test case

**Comment by tshradheya** [comment]
Can you resolve the conflicts?

Might have to change the `expected` values of assertion checks

**Comment by jacoblipech** [comment]
@tshradheya Ready for review!

**Comment by jacoblipech** [comment]
@xpdavid Do you mind helping me with this review? :)

File: src/main/java/teammates/logic/core/FeedbackResponsesLogic.java [hunk 6]
```diff
@@ -91,6 +91,14 @@ public FeedbackResponseAttributes getFeedbackResponse(
         return frDb.getFeedbackResponsesForSessionInSection(feedbackSessionName, courseId, section);
     }
 
+    public List<FeedbackResponseAttributes> getFeedbackResponsesForSessionInGiverAndRecipientSection(
```
Comment (xpdavid) [inline src/main/java/teammates/logic/core/FeedbackResponsesLogic.java]: javadoc?

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 7]
```diff
@@ -1707,6 +1671,43 @@ private FeedbackSessionResultsBundle getFeedbackSessionResultsForUserInSectionBy
...
+        // this happens when 'all' section is selected
+        if (sectionDetail == null) {
+            // default for 'all' section selected
+            params.put(PARAM_IN_SECTION, "true");
+            params.put(PARAM_FROM_SECTION, "false");
+            params.put(PARAM_TO_SECTION, "false");
+            return params;
+        }
+        // either giver or recipient needs to be in the selected section
+        if (sectionDetail.contains("either")) {
```
Comment (xpdavid) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: Should we define `sectionDetails` as an enum?

Comment (jacoblipech) [reply src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: Right now, sectionDetail will be a long text from the selected option using `params.get(PARAM_SECTION_DETAIL)`. By using enum, I can map the long text to keyword based but it will be additional process. 
So I am not sure if it is necessary?

Comment (xpdavid) [reply src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: I kind of have some feelings:

- It would be hard to maintain and understand: It is hard to infer what the `sectionDetails` is as it is just a `String` (and you also allow the possibility of `null`). We can mitigate this by writing javadoc, but I still think `enum` could be a better option as developers can just navigate to the enum.

- It would be hard to add options in the futher: What if you have a option that `contains` words of another options. Are you going to have nested `if`?

- By right, `sectionDetails` is not a variable but rather a option. It should associate with a fixed set of value.

Comment (jacoblipech) [reply src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: Hmm I see. In terms of maintainability, enum seems better. Working on it!

**Comment by jacoblipech** [comment]
@xpdavid Sorry for the delay, ready for review! :)

File: src/main/java/teammates/logic/api/Logic.java [hunk 8]
```diff
@@ -1393,19 +1393,20 @@ public String getFeedbackSessionResultSummaryAsCsv(
      * Preconditions: <br>
      * * All parameters(except questionId) are non-null. <br>
      * @see FeedbackSessionsLogic#getFeedbackSessionResultsSummaryInSectionAsCsv(String, String, String,
-     *      String, String, boolean, boolean)
+     *      String, String, String, boolean, boolean)
      */
     public String getFeedbackSessionResultSummaryInSectionAsCsv(
             String courseId, String feedbackSessionName, String instructorEmail,
-            String section, String questionId, boolean isMissingResponsesShown, boolean isStatsShown)
+            String section, String sectionDetail, String questionId, boolean isMissingResponsesShown, boolean isStatsShown)
```
Comment (xpdavid) [inline src/main/java/teammates/logic/api/Logic.java]: `sectionDetails` should be type of the enum you defined right?

File: src/main/java/teammates/logic/core/FeedbackResponsesLogic.java [hunk 9]
```diff
@@ -91,6 +91,22 @@ public FeedbackResponseAttributes getFeedbackResponse(
         return frDb.getFeedbackResponsesForSessionInSection(feedbackSessionName, courseId, section);
     }
 
+    /**
+     * Retrieves the list of feedback responses for the section selected.
+     * @param feedbackSessionName       Feedback session name
```
Comment (xpdavid) [inline src/main/java/teammates/logic/core/FeedbackResponsesLogic.java]: Should have a blank line between the first sentence and `@param`.

File: src/main/java/teammates/logic/core/FeedbackResponsesLogic.java [hunk 10]
```diff
@@ -91,6 +91,22 @@ public FeedbackResponseAttributes getFeedbackResponse(
...
+     * @param courseId                  Course ID
+     * @param section                   Name of the section selected
+     * @return                          List of feedback response attributes that matches the section selected
+     */
+
```
Comment (xpdavid) [inline src/main/java/teammates/logic/core/FeedbackResponsesLogic.java]: Should no have the extra empty line right?

File: src/main/java/teammates/logic/core/FeedbackResponsesLogic.java [hunk 11]
```diff
@@ -91,6 +91,22 @@ public FeedbackResponseAttributes getFeedbackResponse(
         return frDb.getFeedbackResponsesForSessionInSection(feedbackSessionName, courseId, section);
     }
 
+    /**
+     * Retrieves the list of feedback responses for the section selected.
+     * @param feedbackSessionName       Feedback session name
+     * @param courseId                  Course ID
+     * @param section                   Name of the section selected
```
Comment (xpdavid) [inline src/main/java/teammates/logic/core/FeedbackResponsesLogic.java]: If you all section to be null, you should say here.

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageAction.java [hunk 12]
```diff
@@ -40,9 +41,13 @@ protected ActionResult execute() throws EntityDoesNotExistException {
 
         InstructorFeedbackResultsPageData data = new InstructorFeedbackResultsPageData(account, sessionToken);
         String selectedSection = getRequestParamValue(Const.ParamsNames.FEEDBACK_RESULTS_GROUPBYSECTION);
+        SectionDetail selectedSectionDetail = null;
+        String sectionDetailValue = getRequestParamValue(Const.ParamsNames.FEEDBACK_RESULTS_GROUPBYSECTIONDETAIL);
 
         if (selectedSection == null) {
             selectedSection = ALL_SECTION_OPTION;
+        } else if (sectionDetailValue != null && !sectionDetailValue.isEmpty()) {
```
Comment (xpdavid) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageAction.java]: Under what situation will sectionDetailsValue be `null` or empty?

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageAction.java [hunk 13]
```diff
@@ -40,9 +41,13 @@ protected ActionResult execute() throws EntityDoesNotExistException {
 
         InstructorFeedbackResultsPageData data = new InstructorFeedbackResultsPageData(account, sessionToken);
         String selectedSection = getRequestParamValue(Const.ParamsNames.FEEDBACK_RESULTS_GROUPBYSECTION);
+        SectionDetail selectedSectionDetail = null;
```
Comment (xpdavid) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageAction.java]: What is the meaning of null of SectionDetails?

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageAction.java [hunk 14]
```diff
@@ -192,10 +199,11 @@ private FeedbackSessionResultsBundle getBundleForQuestionView(
...
                 bundle.isComplete = true;
             } else {
-                // bundle for all questions, with a selected section
+                // bundle for all questions, with a selected section and selected section detail
+                // depending on the section detail, this function will filter accordingly
                 bundle = logic.getFeedbackSessionResultsForInstructorInSection(feedbackSessionName, courseId,
-                                                                                    instructor.email,
-                                                                                    selectedSection);
+                                                                                instructor.email, selectedSection,
+                                                                                selectedSectionDetail.name());
```
Comment (xpdavid) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageAction.java]: Why pass the name here?

File: src/main/java/teammates/storage/api/FeedbackResponsesDb.java [hunk 15]
```diff
@@ -473,25 +484,50 @@ private FeedbackResponse getFeedbackResponseEntity(
...
+                String feedbackQuestionId, String section, String sectionDetail) {
         List<FeedbackResponse> feedbackResponses = new ArrayList<>();
+        if ("BOTH".equals(sectionDetail)) {
+            feedbackResponses.addAll(load()
+                    .filter("feedbackQuestionId =", feedbackQuestionId)
+                    .filter("giverSection =", section)
+                    .filter("receiverSection =", section)
+                    .list());
+
+            // show responses in section with giver but without recipient
```
Comment (xpdavid) [inline src/main/java/teammates/storage/api/FeedbackResponsesDb.java]: Why in "BOTH" means "None" should be included?

File: src/main/java/teammates/storage/api/FeedbackResponsesDb.java [hunk 16]
```diff
@@ -473,25 +484,50 @@ private FeedbackResponse getFeedbackResponseEntity(
     }
 
     private List<FeedbackResponse> getFeedbackResponseEntitiesForQuestionInSection(
-                String feedbackQuestionId, String section) {
+                String feedbackQuestionId, String section, String sectionDetail) {
         List<FeedbackResponse> feedbackResponses = new ArrayList<>();
```
Comment (xpdavid) [inline src/main/java/teammates/storage/api/FeedbackResponsesDb.java]: Consider use HashSet to remove duplications.

File: src/main/java/teammates/logic/api/Logic.java (hunk 8 above)
Comment (jacoblipech) [reply src/main/java/teammates/logic/api/Logic.java]: I used string instead as the `FeedbackSessionsLogic.java` has multiple usage of sectionDetail in terms of String as it is put into a `Map<String, String> params` so I cannot put in an enum instead. Hence, I decided to change it to String for `Logic` onwards instead of changing all the existing implementations.

Comment (xpdavid) [reply src/main/java/teammates/logic/api/Logic.java]: it is put into a Map<String, String> params so I cannot put in an enum instead.

Using String will loss the point of using enum. Let's see if we can find any workaround.

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageAction.java (hunk 12 above)
Comment (jacoblipech) [reply src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageAction.java]: sectionDetailsValue is null when the `selectedSection` is all. When a session result is shown without the option to choose a filter section, the sectionDetailsValue is empty.

Comment (xpdavid) [reply src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageAction.java]: Maybe you want to use a special value in the enum instead of making it `null`.

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageAction.java (hunk 13 above)
Comment (jacoblipech) [reply src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageAction.java]: This is to initialise the selectedSectionDetail to be able to be called in the later stage of the code

File: src/main/java/teammates/storage/api/FeedbackResponsesDb.java (hunk 15 above)
Comment (jacoblipech) [reply src/main/java/teammates/storage/api/FeedbackResponsesDb.java]: I followed the original implementation before this PR. For cases when it is directed to a team, the section selected will be "none". Somewhat, "none" can belong to any section.

Since "none" has no specific section, should I change it to implement for all cases instead?

File: src/main/java/teammates/logic/api/Logic.java (hunk 8 above)
Comment (jacoblipech) [reply src/main/java/teammates/logic/api/Logic.java]: I can pass in a separate `SectionDetail` parameter to the functions if the selectedSectionDetail is not `null`?

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageAction.java (hunk 12 above)
Comment (jacoblipech) [reply src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageAction.java]: In this case, do you suggest a "DEFAULT" value in enum then?

File: src/main/java/teammates/storage/api/FeedbackResponsesDb.java (hunk 15 above)
Comment (xpdavid) [reply src/main/java/teammates/storage/api/FeedbackResponsesDb.java]: Before this PR, when section is selected, we only show response which has `giverSection = section` right?

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageAction.java (hunk 12 above)
Comment (xpdavid) [reply src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageAction.java]: If `ALL_SECTION_OPTION` is selected and you decide to ignore the value, I believe `NOT_APPLICABLE` is more appropriate.

File: src/main/java/teammates/storage/api/FeedbackResponsesDb.java (hunk 15 above)
Comment (jacoblipech) [reply src/main/java/teammates/storage/api/FeedbackResponsesDb.java]: yes that is true. my bad... So should I add "none" to all filters of sectionDetails?

**Comment by jacoblipech** [comment]
@xpdavid Ready for review! :) Sorry for the long delay! Hoping to merge this ASAP as I modified parts of `typicalDataBundle` which introduced dependencies from other PRs... 

The main part is to change student2 in course 1 from section1 team1.1 to section2 team1.2. I also change the response and comment from S2C2 to S1C3. (This is to change the student to section 2 for testing of filtering of response and comments from selected section.)

**Comment by xpdavid** [comment]
> as I modified parts of typicalDataBundle which introduced dependencies from other PRs...

What's the impact of the change? If it is large, you can actually make a copy of `typicalDataBundle.json`.

**Comment by jacoblipech** [comment]
>What's the impact of the change? If it is large, you can actually make a copy of typicalDataBundle.json.

I would say that it is not that major to do that .__. For PRs related to changes in the format of feedbackSession responses and addition of new tests in the area, the feedback results tests which I added in this PR need to change accordingly. Mainly this PR has been open for quite some time and I hope to merge it soon :)

**Comment by jacoblipech** [comment]
@xpdavid reminder to review :)

**Comment by jacoblipech** [comment]
@xpdavid Can help to review this? I reworked the testing area for this PR to make it neater.

**Comment by wkurniawan07** [comment]
@xpdavid allow me to take over reviewing this because I will be migrating this page later on.

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsDownloadAction.java [hunk 17]
```diff
@@ -35,6 +38,11 @@ protected ActionResult execute() throws EntityDoesNotExistException {
         String fileContent;
         String fileName;
 
+        // initialize SectionDetail correctly
+        if (section != null && sectionDetailValue != null && !sectionDetailValue.isEmpty()) {
+            sectionDetail = SectionDetail.valueOf(sectionDetailValue);
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsDownloadAction.java]: Need to add a guard if `sectionDetailValue` is invalid.

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageAction.java [hunk 18]
```diff
@@ -39,9 +40,13 @@ protected ActionResult execute() throws EntityDoesNotExistException {
 
         InstructorFeedbackResultsPageData data = new InstructorFeedbackResultsPageData(account, sessionToken);
         String selectedSection = getRequestParamValue(Const.ParamsNames.FEEDBACK_RESULTS_GROUPBYSECTION);
+        String sectionDetailValue = getRequestParamValue(Const.ParamsNames.FEEDBACK_RESULTS_GROUPBYSECTIONDETAIL);
+        SectionDetail selectedSectionDetail = SectionDetail.NOT_APPLICABLE;
 
         if (selectedSection == null) {
             selectedSection = ALL_SECTION_OPTION;
+        } else if (sectionDetailValue != null && !sectionDetailValue.isEmpty()) {
+            selectedSectionDetail = SectionDetail.valueOf(sectionDetailValue);
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageAction.java]: Similarly here; need a guard.

File: src/test/java/teammates/test/cases/logic/FeedbackResponsesLogicTest.java [hunk 19]
```diff
@@ -432,33 +433,36 @@ private void testGetViewableResponsesForQuestionInSection() throws Exception {
...
         List<FeedbackResponseAttributes> responses =
                 frLogic.getViewableFeedbackResponsesForQuestionInSection(fq, instructor.email,
-                                                                         UserRole.INSTRUCTOR, null);
+                                                                         UserRole.INSTRUCTOR, null, null);
 
         assertEquals(1, responses.size());
 
         ______TS("success: GetViewableResponsesForQuestionInSection - instructor");
 
+        // other more in-dept sectionDetail types are tested in FeedbackResponsesDbTest.java
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/logic/FeedbackResponsesLogicTest.java]: in-depth (typo)

File: src/test/java/teammates/test/cases/logic/StudentsLogicTest.java [hunk 20]
```diff
@@ -1025,12 +1025,6 @@ private void testIsStudentsInSameTeam() {
 
         ______TS("students of different teams");
 
-        StudentAttributes student5InCourse1 = dataBundle.students.get("student5InCourse1");
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/logic/StudentsLogicTest.java]: Why is this removed?

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsDownloadAction.java (hunk 17 above)
Comment (jacoblipech) [reply src/main/java/teammates/ui/controller/InstructorFeedbackResultsDownloadAction.java]: Guard as invalid meaning? Currently `selectedSectionDetail` will be default if the field `sectionDetailValue` is null. Do you mean checking if the `sectionDetailValue` has the same value as one of the enum shown in `SectionDetail.java`?

File: src/test/java/teammates/test/cases/logic/StudentsLogicTest.java (hunk 20 above)
Comment (jacoblipech) [reply src/test/java/teammates/test/cases/logic/StudentsLogicTest.java]: Sorry about this. I removed thinking it is the same tests. Will add it back

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsDownloadAction.java (hunk 17 above)
Comment (wkurniawan07) [reply src/main/java/teammates/ui/controller/InstructorFeedbackResultsDownloadAction.java]: > Do you mean checking if the `sectionDetailValue` has the same value as one of the enum shown in `SectionDetail.java`?

This is what I meant.

**Comment by jacoblipech** [comment]
@wkurniawan07 Ready for review!
